pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/TableConfigUtils.java

```diff
@@ -0,0 +1,831 @@
+   * 1. Validation config
+   * 2. IngestionConfig
+   * 3. TierConfigs
+   * 4. Indexing config
+   * 5. Field Config List
+   *
+   * TODO: Add more validations for each section (e.g. validate conditions are met for aggregateMetrics)
+   */
+  public static void validate(TableConfig tableConfig, @Nullable Schema schema) {
+    if (tableConfig.getTableType() == TableType.REALTIME) {
+      Preconditions.checkState(schema != null, "Schema should not be null for REALTIME table");
+    }
+    // Sanitize the table config before validation
+    sanitize(tableConfig);
+    validateValidationConfig(tableConfig, schema);
+    validateIngestionConfig(tableConfig, schema);
+    validateTierConfigList(tableConfig.getTierConfigsList());
+    validateIndexingConfig(tableConfig.getIndexingConfig(), schema);
+    validateFieldConfigList(tableConfig.getFieldConfigList(), tableConfig.getIndexingConfig(), schema);
+    validateUpsertConfig(tableConfig, schema);
+    validatePartialUpsertStrategies(tableConfig, schema);
+    validateTaskConfigs(tableConfig, schema);
+  }
+
+  /**
+   * Validates the table name with the following rules:
+   * <ul>
+   *   <li>Table name shouldn't contain dot or space in it</li>
+   * </ul>
+   */
+  public static void validateTableName(TableConfig tableConfig) {
+    String tableName = tableConfig.getTableName();
+    if (tableName.contains(".") || tableName.contains(" ")) {
+      throw new IllegalStateException("Table name: '" + tableName + "' containing '.' or space is not allowed");
+    }
+  }
+
+  /**
+   * Validates retention config. Checks for following things:
+   * - Valid segmentPushType
+   * - Valid retentionTimeUnit
+   */
+  private static void validateRetentionConfig(TableConfig tableConfig) {
+    SegmentsValidationAndRetentionConfig segmentsConfig = tableConfig.getValidationConfig();
+    String tableName = tableConfig.getTableName();
+
+    if (segmentsConfig == null) {
+      throw new IllegalStateException(
+          String.format("Table: %s, \"segmentsConfig\" field is missing in table config", tableName));
+    }
+
+    String segmentPushType = IngestionConfigUtils.getBatchSegmentIngestionType(tableConfig);
+    // segmentPushType is not needed for Realtime table
+    if (tableConfig.getTableType() == TableType.OFFLINE && segmentPushType != null && !segmentPushType.isEmpty()) {
+      if (!segmentPushType.equalsIgnoreCase("REFRESH") && !segmentPushType.equalsIgnoreCase("APPEND")) {
+        throw new IllegalStateException(String.format("Table: %s, invalid push type: %s", tableName, segmentPushType));
+      }
+    }
+
+    // Retention may not be specified. Ignore validation in that case.
+    String timeUnitString = segmentsConfig.getRetentionTimeUnit();
+    if (timeUnitString == null || timeUnitString.isEmpty()) {
+      return;
+    }
+    try {
+      TimeUnit.valueOf(timeUnitString.toUpperCase());
+    } catch (Exception e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)